In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import yfinance as yf
import datetime as dt
from pandas.tseries.offsets import MonthEnd

In [24]:
# getting table elements in wikipedia 
table = pd.read_html('https://en.wikipedia.org/wiki/List_of_S%26P_500_companies')[0]


In [31]:
tickers = table.Symbol.to_list()

In [33]:
len(tickers)

504

In [35]:
df = yf.download(tickers, start='2009-01-01', end='2022-01-01')['Adj Close']

[*********************100%***********************]  504 of 504 completed

3 Failed downloads:
- BRK.B: No data found, symbol may be delisted
- CEG: Data doesn't exist for startDate = 1230778800, endDate = 1641006000
- BF.B: No data found for this date range, symbol may be delisted
[*********************102%************************]  515 of 504 completed

In [37]:
df.index = pd.to_datetime(df.index)

In [39]:
monthly_ret = df.pct_change().resample('M').agg(lambda x: (x+1).prod() -1)

[*************************116%***************************]  587 of 504 completed

In [40]:
monthly_ret

A       AAL       AAP      AAPL      ABBV       ABC  \
Date                                                                     
2008-12-31  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2009-01-31  0.113301 -0.324196 -0.041301 -0.006832  0.000000  0.021085   
2009-02-28 -0.232854 -0.497355  0.168653 -0.009098  0.000000 -0.123204   
2009-03-31  0.108147 -0.112281  0.075635  0.177023  0.000000  0.028337   
2009-04-30  0.188028  0.498023  0.064995  0.197013  0.000000  0.030006   
...              ...       ...       ...       ...       ...       ...   
2021-08-31  0.145141 -0.021590 -0.043431  0.042489  0.038521  0.004008   
2021-09-30 -0.102240  0.029087  0.034794 -0.068037 -0.106888 -0.022584   
2021-10-31  0.000991 -0.064327  0.079611  0.058657  0.075926  0.021515   
2021-11-30 -0.041844 -0.078646 -0.021284  0.105082  0.005320 -0.047901   
2021-12-31  0.057985  0.015263  0.091371  0.074229  0.174532  0.148078   

                ABMD       ABT       ACN      ADBE  ...      WYNN       XEL  \
Date                                                ...                       
2008-12-31  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   
2009-01-31 -0.173913  0.042599 -0.062667 -0.161164  ... -0.365668 -0.020170   
2009-02-28 -0.495182 -0.146104 -0.075095 -0.135163  ... -0.303524 -0.039003   
2009-03-31 -0.280470  0.007604 -0.058239  0.280838  ... -0.046778  0.063656   
2009-04-30  0.361224 -0.114598  0.070571  0.278635  ...  0.964446 -0.010198   
...              ...       ...       ...       ...  ...       ...       ...   
2021-08-31  0.112551  0.044553  0.059431  0.067677  ...  0.034171  0.007326   
2021-09-30 -0.105616 -0.065205 -0.049441 -0.132560  ... -0.166585 -0.084660   
2021-10-31  0.020030  0.095299  0.124829  0.129646  ...  0.059587  0.033440   
2021-11-30 -0.051982 -0.024207 -0.003874  0.029968  ... -0.097884 -0.013315   
2021-12-31  0.141019  0.119027  0.159905 -0.153452  ...  0.049747  0.069593   

                 XOM      XRAY       XYL       YUM       ZBH      ZBRA  \
Date                                                                     
2008-12-31  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2009-01-31 -0.063204 -0.075576  0.000000 -0.099904 -0.128353 -0.199715   
2009-02-28 -0.107712 -0.140840  0.000000 -0.081760 -0.037912  0.043969   
2009-03-31  0.002946  0.163565  0.000000  0.045662  0.042262  0.082527   
2009-04-30 -0.020999  0.065921  0.000000  0.221385  0.205206  0.117245   
...              ...       ...       ...       ...       ...       ...   
2021-08-31 -0.038645 -0.065718  0.085383  0.000983 -0.079366  0.062790   
2021-09-30  0.078870 -0.057422 -0.092656 -0.066550 -0.025635 -0.122196   
2021-10-31  0.096056 -0.014470  0.055870  0.021503 -0.022137  0.035951   
2021-11-30 -0.059342 -0.148051 -0.070578 -0.012909 -0.164338  0.102688   
2021-12-31  0.022560  0.146872 -0.009826  0.130414  0.064219  0.010904   

                ZION       ZTS  
Date                            
2008-12-31  0.000000  0.000000  
2009-01-31 -0.403200  0.000000  
2009-02-28 -0.370227  0.000000  
2009-03-31  0.049093  0.000000  
2009-04-30  0.111902  0.000000  
...              ...       ...  
2021-08-31  0.117938  0.009176  
2021-09-30  0.068912 -0.050939  
2021-10-31  0.017773  0.114959  
2021-11-30  0.007285  0.027012  
2021-12-31  0.001268  0.099036  

[157 rows x 504 columns]

[***************************124%*****************************]  624 of 504 completed

In [42]:
#build past 11 month returns 
# the +1 and -1 is for multiplying percentages
past_11 = (monthly_ret+1).rolling(11).apply(np.prod)-1

In [45]:
past_11 

A       AAL       AAP      AAPL      ABBV       ABC  \
Date                                                                     
2008-12-31       NaN       NaN       NaN       NaN       NaN       NaN   
2009-01-31       NaN       NaN       NaN       NaN       NaN       NaN   
2009-02-28       NaN       NaN       NaN       NaN       NaN       NaN   
2009-03-31       NaN       NaN       NaN       NaN       NaN       NaN   
2009-04-30       NaN       NaN       NaN       NaN       NaN       NaN   
...              ...       ...       ...       ...       ...       ...   
2021-08-31  0.749301  0.622457  0.332136  0.319459  0.447199  0.280847   
2021-09-30  0.549967  0.819149  0.436702  0.308195  0.312456  0.262997   
2021-10-31  0.354944  0.358811  0.546670  0.264207  0.149074  0.197097   
2021-11-30  0.280827  0.121750  0.417271  0.253441  0.127480  0.202130   
2021-12-31  0.333964  0.046010  0.633575  0.353931  0.368600  0.294840   

                ABMD       ABT       ACN      ADBE  ...      WYNN       XEL  \
Date                                                ...                       
2008-12-31       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2009-01-31       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2009-02-28       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2009-03-31       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2009-04-30       NaN       NaN       NaN       NaN  ...       NaN       NaN   
...              ...       ...       ...       ...  ...       ...       ...   
2021-08-31  0.313650  0.178497  0.509101  0.353302  ...  0.416098  0.017005   
2021-09-30  0.292361  0.136853  0.488708  0.287676  ...  0.170095 -0.082653   
2021-10-31  0.211383  0.209409  0.458207  0.359249  ... -0.106468 -0.014399   
2021-11-30 -0.029056  0.166445  0.385160  0.339379  ... -0.282017 -0.023919   
2021-12-31  0.031357  0.151728  0.728856  0.236044  ... -0.145584  0.087734   

                 XOM      XRAY       XYL       YUM       ZBH      ZBRA  \
Date                                                                     
2008-12-31       NaN       NaN       NaN       NaN       NaN       NaN   
2009-01-31       NaN       NaN       NaN       NaN       NaN       NaN   
2009-02-28       NaN       NaN       NaN       NaN       NaN       NaN   
2009-03-31       NaN       NaN       NaN       NaN       NaN       NaN   
2009-04-30       NaN       NaN       NaN       NaN       NaN       NaN   
...              ...       ...       ...       ...       ...       ...   
2021-08-31  0.703547  0.418441  0.636984  0.459987  0.110202  1.325794   
2021-09-30  0.934253  0.238963  0.433831  0.333183  0.114825  0.817162   
2021-10-31  0.770963  0.132258  0.370743  0.195960 -0.034280  0.410998   
2021-11-30  0.540990 -0.064265  0.201156  0.150508 -0.220271  0.531965   
2021-12-31  0.448542  0.050497  0.253384  0.391148 -0.167930  0.534693   

                ZION       ZTS  
Date                            
2008-12-31       NaN       NaN  
2009-01-31       NaN       NaN  
2009-02-28       NaN       NaN  
2009-03-31       NaN       NaN  
2009-04-30       NaN       NaN  
...              ...       ...  
2021-08-31  1.037776  0.243871  
2021-09-30  0.972330  0.229750  
2021-10-31  0.664210  0.355476  
2021-11-30  0.489174  0.349024  
2021-12-31  0.467416  0.588298  

[157 rows x 504 columns]

[********************************************197%***********************************************]  995 of 504 completed

In [46]:
formation = dt.datetime(2009,12,31)

In [47]:
formation

datetime.datetime(2009, 12, 31, 0, 0)

In [52]:
# we cutoff the recent month before get the total result of the positions
end_measurement = formation - MonthEnd(1)

In [53]:
end_measurement

Timestamp('2009-11-30 00:00:00')

In [54]:
ret_12 = past_11.loc[end_measurement]

In [55]:
ret_12

A       0.780789
AAL    -0.560191
AAP     0.156284
AAPL    1.202866
ABBV    0.000000
          ...   
YUM     0.129019
ZBH     0.416906
ZBRA    0.265335
ZION   -0.470798
ZTS     0.000000
Name: 2009-11-30 00:00:00, Length: 504, dtype: float64

In [56]:
ret_12 = ret_12.reset_index()

In [57]:
ret_12

index  2009-11-30 00:00:00
0       A             0.780789
1     AAL            -0.560191
2     AAP             0.156284
3    AAPL             1.202866
4    ABBV             0.000000
..    ...                  ...
499   YUM             0.129019
500   ZBH             0.416906
501  ZBRA             0.265335
502  ZION            -0.470798
503   ZTS             0.000000

[504 rows x 2 columns]

In [58]:
ret_12['decile'] = pd.qcut(ret_12.iloc[:,1],10,labels=False, duplicates='drop')

In [59]:
ret_12

index  2009-11-30 00:00:00  decile
0       A             0.780789       9
1     AAL            -0.560191       0
2     AAP             0.156284       4
3    AAPL             1.202866       9
4    ABBV             0.000000       1
..    ...                  ...     ...
499   YUM             0.129019       4
500   ZBH             0.416906       7
501  ZBRA             0.265335       6
502  ZION            -0.470798       0
503   ZTS             0.000000       1

[504 rows x 3 columns]

In [60]:
winners = ret_12[ret_12.decile == 9]
losers = ret_12[ret_12.decile == 0]

In [62]:
# searching the winners symbols in monthly returns df 
winners_ret = monthly_ret.loc[formation + MonthEnd(1), monthly_ret.columns.isin(winners['index'])]
losers_ret = monthly_ret.loc[formation + MonthEnd(1), monthly_ret.columns.isin(losers['index'])]

In [64]:
Momentumprofit = winners_ret.mean() - losers_ret.mean()
Momentumprofit

-0.06269413444105827

In [68]:
def momentum(formation):
    end_measurement = formation - MonthEnd(1)
    ret_12 = past_11.loc[end_measurement]
    ret_12 = ret_12.reset_index()
    ret_12['decile'] = pd.qcut(ret_12.iloc[:,1], 10 , labels=False, duplicates='drop')
    winners = ret_12[ret_12.decile == 9]
    losers = ret_12[ret_12.decile == 0]
    winners_ret = monthly_ret.loc[formation + MonthEnd(1), monthly_ret.columns.isin(winners['index'])]
    losers_ret = monthly_ret.loc[formation + MonthEnd(1), monthly_ret.columns.isin(losers['index'])]
    Momentumprofit = winners_ret.mean() - losers_ret.mean()
    return Momentumprofit

In [69]:
momentum(formation)

-0.06269413444105827

In [71]:
formation

datetime.datetime(2009, 12, 31, 0, 0)

In [72]:
for i in range(12*12):
    print(formation + MonthEnd(i))

2009-12-31 00:00:00
2010-01-31 00:00:00
2010-02-28 00:00:00
2010-03-31 00:00:00
2010-04-30 00:00:00
2010-05-31 00:00:00
2010-06-30 00:00:00
2010-07-31 00:00:00
2010-08-31 00:00:00
2010-09-30 00:00:00
2010-10-31 00:00:00
2010-11-30 00:00:00
2010-12-31 00:00:00
2011-01-31 00:00:00
2011-02-28 00:00:00
2011-03-31 00:00:00
2011-04-30 00:00:00
2011-05-31 00:00:00
2011-06-30 00:00:00
2011-07-31 00:00:00
2011-08-31 00:00:00
2011-09-30 00:00:00
2011-10-31 00:00:00
2011-11-30 00:00:00
2011-12-31 00:00:00
2012-01-31 00:00:00
2012-02-29 00:00:00
2012-03-31 00:00:00
2012-04-30 00:00:00
2012-05-31 00:00:00
2012-06-30 00:00:00
2012-07-31 00:00:00
2012-08-31 00:00:00
2012-09-30 00:00:00
2012-10-31 00:00:00
2012-11-30 00:00:00
2012-12-31 00:00:00
2013-01-31 00:00:00
2013-02-28 00:00:00
2013-03-31 00:00:00
2013-04-30 00:00:00
2013-05-31 00:00:00
2013-06-30 00:00:00
2013-07-31 00:00:00
2013-08-31 00:00:00
2013-09-30 00:00:00
2013-10-31 00:00:00
2013-11-30 00:00:00
2013-12-31 00:00:00
2014-01-31 00:00:00


In [74]:
profits = []
dates = []

for i in range(12*12):
    profits.append(momentum(formation + MonthEnd(i)))
    dates.append(formation + MonthEnd(i))

In [75]:
profits

[-0.06269413444105827,
 0.04589333433302381,
 0.11711915472622146,
 0.050206152816435354,
 -0.058388777828795244,
 -0.07411424472305742,
 nan,
 -0.004136842246748894,
 nan,
 nan,
 0.046061469212687466,
 -0.05828239947041519,
 -0.03717741988305373,
 nan,
 0.021309558459286823,
 -0.018186327763180462,
 -0.0031871319818563533,
 0.020882209949230634,
 -0.009499860726427604,
 -0.02760753672861188,
 -0.07152293208656926,
 -0.06759053825317934,
 0.04902672612066786,
 -0.004223342569866964,
 -0.11211739167960164,
 -0.016742060401234313,
 0.029483014747147474,
 0.06765787693492266,
 0.023919466682058986,
 -0.019899720055333782,
 0.012945816554773512,
 -0.012686091041394516,
 -0.011595116787674921,
 -0.001008389425327826,
 0.025629880284345416,
 -0.004367219292935268,
 -0.012110941777297768,
 -0.007242865193899835,
 0.011854446008829463,
 -0.016036711450117203,
 0.01583438092403193,
 -0.01988619868750855,
 0.053027369946046995,
 0.005077588607659073,
 0.061910021840055604,
 0.031726413591601714,

In [77]:
SP = yf.download('^GSPC', start=dates[0], end=dates[-1])['Adj Close']

[*********************100%***********************]  1 of 1 completed


In [78]:
SP_monthly = SP.pct_change().resample('M').agg(lambda x: (x+1).prod() -1)

In [79]:
SP_monthly

Date
2009-12-31    0.000000
2010-01-31   -0.036974
2010-02-28    0.028514
2010-03-31    0.058796
2010-04-30    0.014759
                ...   
2021-07-31    0.022748
2021-08-31    0.028990
2021-09-30   -0.047569
2021-10-31    0.069144
2021-11-30    0.010833
Freq: M, Name: Adj Close, Length: 144, dtype: float64

In [80]:
final = pd.DataFrame(profits)

In [81]:
final['S_P500'] = SP_monthly.values

In [83]:
final['excess'] = final.iloc[:,0] - final.iloc[:,1]

In [85]:
# list comprehension 
final['outperformed'] = ['Yes' if i > 0 else 'No' for i in final.excess]

In [86]:
final

0    S_P500    excess outperformed
0   -0.062694  0.000000 -0.062694           No
1    0.045893 -0.036974  0.082868          Yes
2    0.117119  0.028514  0.088605          Yes
3    0.050206  0.058796 -0.008590           No
4   -0.058389  0.014759 -0.073148           No
..        ...       ...       ...          ...
139  0.024512  0.022748  0.001764          Yes
140  0.007877  0.028990 -0.021114           No
141  0.061609 -0.047569  0.109178          Yes
142 -0.009853  0.069144 -0.078997           No
143 -0.073136  0.010833 -0.083969           No

[144 rows x 4 columns]

In [88]:
final[final.outperformed == 'Yes'].shape

(55, 4)

In [90]:
final.shape

(144, 4)

In [92]:
# winning rate
55/144

0.3819444444444444

# Conclusion
* We get a 38.19% aproximatelly winning rate
* I used some references to build and test this strategy 